### PR TITLE
[SYCL] Cleanup/refactor device_code_tests

### PR DESCRIPTION
**Pull request by ianayl**
This PR:
- Changes more lit tests in `sycl/test/check_device_code` to use `SYCL_EXTERNAL` functions in order to remove redundant host-side boilerplate code, and
- Moves tests not verifying device code outside of the `check_device_code` folder.
- Moves tests verifying device code that are outside of the  `check_device_code` folder into the `check_device_code` folder

### CODE REVIEW

File: sycl/test/check_device_code/ap_fixed.cpp [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// RUN: %clangxx -I %sycl_include -S -emit-llvm -fsycl-device-only %s -o - -Xclang -disable-llvm-passes -Xclang -no-enable-noundef-analysis | FileCheck %s
+// RUN: %clangxx -I %sycl_include -S -emit-llvm -fno-sycl-early-optimizations -fsycl-device-only %s -o - -Xclang -disable-llvm-passes -Xclang -no-enable-noundef-analysis | FileCheck %s
```
Comment (aelovikov-intel) [inline sycl/test/check_device_code/ap_fixed.cpp:1]: Why do we need `-fno-sycl-early-optimizations`? I'd think we mostly care about what the code is after default optimizations pipeline...

File: sycl/test/check_device_code/ap_fixed.cpp [hunk 2]
```diff
@@ -8,103 +8,59 @@
...
 
-#include "CL/__spirv/spirv_ops.hpp"
+#include <sycl/sycl.hpp>
 
-template <int W, int rW, bool S, int I, int rI>
-void sqrt() {
-  sycl::detail::ap_int<W> a;
-  auto ap_fixed_Sqrt = __spirv_FixedSqrtINTEL<W, rW>(a, S, I, rI);
+SYCL_EXTERNAL void test_sqrt(sycl::detail::ap_int<13> a) {
+  auto ap_fixed_Sqrt = __spirv_FixedSqrtINTEL<13, 5>(a, false, 2, 2);
```
Comment (aelovikov-intel) [inline sycl/test/check_device_code/ap_fixed.cpp:14]:
```suggestion
SYCL_EXTERNAL auto test_sqrt(sycl::detail::ap_int<13> a) {
  return __spirv_FixedSqrtINTEL<13, 5>(a, false, 2, 2);
```

And I think that would resolve my question above as well.

File: sycl/test/check_device_code/device_asan_no_ctor.cpp [hunk 3]
```diff
@@ -11,13 +11,6 @@
...
-    queue.submit(
-        [&](sycl::handler &cgh) { cgh.single_task<class t>([=]() {}); });
-    queue.wait();
-  }
-
-  return 0;
-}
+SYCL_EXTERNAL void nothing() {
+  // Intentionally empty
+}
```
Comment (aelovikov-intel) [inline sycl/test/check_device_code/device_asan_no_ctor.cpp:16]: I think we need to drop changes in this test, at least for this round of changes.

File: sycl/test/check_device_code/device_global_const_eval_use.cpp [hunk 4]
```diff
@@ -80,42 +80,38 @@ constexpr device_global<TestStruct2, decltype(properties(device_image_scope))>
...
-
-      // Struct and array of structs
-      std::ignore = dg_struct.get().field1;
-      std::ignore = dg_struct_arr[0];
-
-      // Struct with constexpr constructor
-      std::ignore = dg_constexpr_constructor_struct.get().value;
-    });
-  });
+SYCL_EXTERNAL void device_global_const_eval_use() {
```
Comment (aelovikov-intel) [inline sycl/test/check_device_code/device_global_const_eval_use.cpp:83]: Do we even need this function? If yes, can we put the attribute on the `device_global` variables instead and remove it still?

File: sycl/test/check_device_code/device_has.cpp [hunk 5]
```diff
@@ -45,29 +44,19 @@ constexpr sycl::aspect getAspect() { return sycl::aspect::cpu; }
...
+  func2();
+  func3();
+  func4<sycl::aspect::host>();
+  func5();
+  func6();
 }
 
+// CHECK-ASPECTS: define dso_local spir_func void @{{.*}}kernel_name_2{{.*}} !sycl_declared_aspects ![[ASPECTS4:[0-9]+]]
+// CHECK-SRCLOC: define dso_local spir_func void @{{.*}}kernel_name_2{{.*}} !srcloc ![[SRCLOC8:[0-9]+]] {{.*}}
+SYCL_EXTERNAL [[sycl::device_has(sycl::aspect::gpu)]] void kernel_name_2() {}
```
Comment (aelovikov-intel) [inline sycl/test/check_device_code/device_has.cpp:58]: I think we might need *both* new and old code, actually. Someone from @intel/dpcpp-tools-reviewers should review changes in this file.

File: sycl/test/check_device_code/fpga_ihs_float.cpp [hunk 6]
```diff
@@ -1,4 +1,5 @@
-//==- fpga_ihs_float.cpp - SYCL FPGA arbitrary precision floating point test -==//
+//==- fpga_ihs_float.cpp - SYCL FPGA arbitrary precision floating point test
+//-==//
```
Comment (aelovikov-intel) [inline sycl/test/check_device_code/fpga_ihs_float.cpp:2]: Drop this change.

File: sycl/test/check_device_code/fpga_ihs_float.cpp [hunk 7]
```diff
@@ -395,55 +368,47 @@ void ap_float_pown() {
...
-    ap_float_atapin<1, 38, 1, 32>();
-    ap_float_atan2<7, 16, 7, 17, 8, 18>();
-    ap_float_pow<8, 8, 9, 9, 10, 10>();
-    ap_float_powr<18, 35, 19, 35, 20, 35>();
-    ap_float_pown<4, 7, 10, 5, 9>();
-  });
-  return 0;
-}
+SYCL_EXTERNAL void fpga_ihs_float() {
+  ap_float_cast<11, 28, 9, 30>();
```
Comment (aelovikov-intel) [inline sycl/test/check_device_code/fpga_ihs_float.cpp:372]: Why not eliminate all the templates similarly to the changes in one of the files above?

File: sycl/test/check_device_code/group_ballot.cpp [hunk 8]
```diff
@@ -3,9 +3,6 @@
 
 #include <sycl/sycl.hpp>
 
-int main() {
-  sycl::queue Q;
-  Q.parallel_for(sycl::nd_range<1>{32, 32}, [=](sycl::nd_item<1> item) {
-    auto Mask = sycl::ext::oneapi::group_ballot(item.get_sub_group());
-  });
+SYCL_EXTERNAL void cast_group_ballots(sycl::nd_item<1> item) {
```
Comment (aelovikov-intel) [inline sycl/test/check_device_code/group_ballot.cpp:6]: This test should also be moved out of this folder.

File: sycl/test/check_device_code/matrix/matrix_load_store_as.cpp [hunk 9]
```diff
@@ -9,55 +9,32 @@
...
-          joint_matrix<sub_group, unsigned short, use::b, 16, 16,
-                       layout::ext_intel_packed>
-              tB;
-          joint_matrix<sub_group, float, use::accumulator, 8, 16> tC;
-
-          sub_group sg = it.get_sub_group();
-          vec<unsigned short, 8> slmvec = sg.load<8>(pA);
-          sg.store<8>(
-              tileA.template get_multi_ptr<sycl::access::decorated::yes>(),
+SYCL_EXTERNAL [[intel::reqd_sub_group_size(16)]] void matrix_store_as(
```
Comment (aelovikov-intel) [inline sycl/test/check_device_code/matrix/matrix_load_store_as.cpp:12]: Matrix tests belong to a different team (or at least should) and thus should go to a separate PR.

File: sycl/test/check_device_code/device_global_const_eval_use.cpp (hunk 4 above)
Comment (ianayl) [reply sycl/test/check_device_code/device_global_const_eval_use.cpp:83]: You'll have to excuse me: Which attribute? SYCL_EXTERNAL? As far as I can tell, SYCL_EXTERNAL cannot be applied to variables in the anonymous/global namespace.

File: sycl/test/check_device_code/group_ballot.cpp (hunk 8 above)
Comment (ianayl) [reply sycl/test/check_device_code/group_ballot.cpp:6]: Should the group tests have their own group folder? Or does it just not belong in check_device_code?

File: sycl/test/check_device_code/matrix/matrix_load_store_as.cpp (hunk 9 above)
Comment (ianayl) [reply sycl/test/check_device_code/matrix/matrix_load_store_as.cpp]: According to the CODEOWNERS file, this is _technically_ ours, although that is probably a typo. Either way, I've reverted the changes, and will be making a new PR for the matrix team.

Should I do anything about CODEOWNERS? I will communicate with the matrix team regardless.

File: sycl/test/check_device_code/device_asan_no_ctor.cpp (hunk 3 above)
Comment (ianayl) [reply sycl/test/check_device_code/device_asan_no_ctor.cpp]: I've reverted the changes, although for my understanding's sake, can I ask why?

Comment (aelovikov-intel) [reply sycl/test/check_device_code/device_asan_no_ctor.cpp]: I'm not sure if the transformation works on device functions or device kernels. If the latter, then the modified test didn't really go through the intended code path.

File: sycl/test/check_device_code/device_global_const_eval_use.cpp (hunk 4 above)
Comment (aelovikov-intel) [reply sycl/test/check_device_code/device_global_const_eval_use.cpp:83]: >  As far as I can tell, SYCL_EXTERNAL cannot be applied to variables in the anonymous/global namespace.

That answers my question, and the change here is good.

File: sycl/test/check_device_code/group_ballot.cpp (hunk 8 above)
Comment (aelovikov-intel) [reply sycl/test/check_device_code/group_ballot.cpp:6]: The latter - LLVM IR isn't verified in any way in this test.

File: sycl/test/check_device_code/matrix/matrix_load_store_as.cpp (hunk 9 above)
Comment (aelovikov-intel) [reply sycl/test/check_device_code/matrix/matrix_load_store_as.cpp]: > Should I do anything about CODEOWNERS? I will communicate with the matrix team regardless.

Yes, please create a PR to update it.

File: sycl/test/check_device_code/group_ballot.cpp (hunk 8 above)
Comment (aelovikov-intel) [reply sycl/test/check_device_code/group_ballot.cpp:6]: I'd start with `sycl/test/extensions/group_ballot.cpp`

File: sycl/test/check_device_code/ap_fixed.cpp (hunk 2 above)
Comment (ianayl) [reply sycl/test/check_device_code/ap_fixed.cpp]: Fixed, thanks for the pointer!

File: sycl/test/check_device_code/fpga_ihs_float.cpp (hunk 6 above)
Comment (ianayl) [reply sycl/test/check_device_code/fpga_ihs_float.cpp]: Fixed!

File: sycl/test/check_device_code/fpga_ihs_float.cpp (hunk 7 above)
Comment (ianayl) [reply sycl/test/check_device_code/fpga_ihs_float.cpp]: My bad, this was one of the earlier test cases I did, I didn't know I could do it as templates. It's fixed now!

File: sycl/test/check_device_code/ap_fixed.cpp [hunk 10]
```diff
@@ -8,103 +8,59 @@
 //
 //===----------------------------------------------------------------------===//
 
-#include "CL/__spirv/spirv_ops.hpp"
+#include <sycl/sycl.hpp>
```
Comment (AlexeySachkov) [inline sycl/test/check_device_code/ap_fixed.cpp:11]: Why do we need the whole `sycl/sycl.hpp` here? Especially if we don't use kernel submission API anymore

File: sycl/test/check_device_code/kernel_arguments_as.cpp [hunk 11]
```diff
@@ -31,34 +31,21 @@
...
-int main() {
-
-  sycl::queue queue;
-  int array[10] = {0};
-  {
-    sycl::buffer<int, 1> buf((int *)array, sycl::range<1>(10),
-                             {sycl::property::buffer::use_host_ptr()});
-    queue.submit([&](sycl::handler &cgh) {
-      auto acc = buf.get_access<sycl::access::mode::read_write>(cgh);
+SYCL_EXTERNAL void check_address_space(
```
Comment (AlexeySachkov) [inline sycl/test/check_device_code/kernel_arguments_as.cpp:38]: This test is specifically about kernel arguments, I'm not sure if rewriting it is correct.

File: sycl/test/check_device_code/ap_fixed.cpp (hunk 10 above)
Comment (ianayl) [reply sycl/test/check_device_code/ap_fixed.cpp]: My bad! My editor was complaining about missing certain SYCL features, so I had assumed that I needed sycl.hpp.

This is fixed now!

File: sycl/test/check_device_code/device_has.cpp (hunk 5 above)
Comment (ianayl) [reply sycl/test/check_device_code/device_has.cpp]: I split the test cases into 2 files, one for kernels and one for functions. Am I good to proceed with this?

File: sycl/test/check_device_code/kernel_arguments_as.cpp (hunk 11 above)
Comment (ianayl) [reply sycl/test/check_device_code/kernel_arguments_as.cpp]: That makes sense, I've gone ahead and reverted the changes.

**Comment by ianayl** [comment]
@AlexeySachkov Please excuse the ping, I've made some changes following your comments: Are these changes good, or do they need further revision?

Thanks in advance!

**Review by AlexeySachkov — APPROVED** [review]

**Comment by ianayl** [comment]
@aelovikov-intel Are these changes good? I believe I still need your approval before I can move onto CI/github workflow. Thanks!

**Comment by aelovikov-intel** [comment]
> @aelovikov-intel Are these changes good? I believe I still need your approval before I can move onto CI/github workflow. Thanks!

These are different approvals. I'll continue my PR review later today.

**Review by aelovikov-intel — APPROVED** [review]

**Comment by ianayl** [comment]
Hey all,

Terribly sorry for the mess here: I was trying to remove certain commits from this branch and didn't know rebasing would cause everyone in a prior merge to be notified. I will be opening up a new branch for this instead.

**Comment by bader** [comment]
@ianayl, please, reset your branch to 88cba5e7b34d5a66bf88a70bc6913a5696794d2d and re-open the pull request.
Avoid using rebase and just put fixes as separate commits on top of your branch.

**Comment by ianayl** [comment]
@bader My bad, that is what I am currently doing. Thanks for your patience!

**Pull request closed without merging**